xtc/src/ledger.rs

```diff
@@ -12,6 +13,9 @@ use std::collections::HashMap;
 pub struct Ledger {
     // stores the cycle balance hold by the principal
     balances: HashMap<Principal, u64>,
+
+    // stores the allowances, approving Principal -> spender principal -> cycle balanace
+    allowances: HashMap<Principal, HashMap<Principal, u64>>,
```

Comment:
Why not use a composit key `HashMap<(Principal, Principal), u64>`?

Reply:
hmmm, I do like that idea, it would make the cleanup code simpler

Reply:
wait I think it is still better to keep the original datastructure as it is more space efficient:

Alice -> Bob1
Alice -> Bob2
Alice -> Bob3
Alice -> Bob4
....
Alice -> Bob10000

in this case we store Alice 10000 times using the datastructure you proposed